Modules/_asynciomodule.c

```diff
@@ -3636,6 +3637,7 @@ _asyncio_all_tasks_impl(PyObject *module, PyObject *loop)
     while (head != tail)
     {
         if (add_one_task(state, tasks, (PyObject *)head, loop) < 0) {
+            ASYNCIO_STATE_UNLOCK(state);
```

Comment:
I'm worried that `add_one_task` might reentrantly call this function (or some other function that uses the state lock), and trigger a deadlock.

We should probably switch `ASYNCIO_STATE_LOCK` to use a critical section -- that would avoid this risk. I think it'll require a bit of refactoring:

* `#define ASYNCIO_STATE_LOCK(state) Py_BEGIN_CRITICAL_SECTION_MUT((&state->mutex)`
* `#define ASYNCIO_STATE_UNLOCK(state) Py_END_CRITICAL_SECTION()`

The usages will have to be refactored a bit because `Py_BEGIN_CRITICAL_SECTION_MUT` creates a new scope in C. (It begins with a `{`).

Reply:
Refactored it a bit to use critical section